File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/userBusinessTiers/UserBusinessTiersHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.userBusinessTiers

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.userBusinessTiers.UserBusinessTiersApi
import features.userBusinessTiers.models.UserProfileWithTierProgram

import features.userBusinessTiers.serializeUserBusinessTiersState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class UserBusinessTiersHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_user_business_tiers"

    override val eventChannel: String
        get() = "customer_sdk/events/get_user_business_tiers"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getUserBusinessTiers(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getUserBusinessTiers(call: MethodCall): String? {
            val result = UserBusinessTiersApi().getUserBusinessTiers(
                memberBusinessId = call.argument<String>("memberBusinessId") as String
            ).toJsonResult<UserProfileWithTierProgram>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        UserBusinessTiersApi().userBusinessTiersState.serializeUserBusinessTiersState().observe { result ->
            pushSerializedResultToEventSink<UserBusinessTiersApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/userBusinessTiers/UserBusinessTiersHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class UserBusinessTiersHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_user_business_tiers"

    var eventChannel: String = "customer_sdk/events/get_user_business_tiers"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await UserBusinessTiersHandler.getUserBusinessTiers(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getUserBusinessTiers(call: FlutterMethodCall) async -> String? {
        

        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
                        guard let memberBusinessId = myArgs["memberBusinessId"] as? String else {return nil}

         let result = try await UserBusinessTiersApi().getUserBusinessTiers(
                memberBusinessId: memberBusinessId
            ).serializeUserBusinessTiersApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        UserBusinessTiersApi().userBusinessTiersState
            .serializeUserBusinessTiersState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
